### ISSUE TITLE
oneTBB as a Swift Package, the Open-Source Metaverse starts with you.

### ISSUE THREAD

**Issue by furby-tm** [oneapi-src/oneTBB#1244]
Hello all,

This is a long-shot but **TBB** has proven to be quite a substantial dependency in the long list of dependencies apart of the **Metaverse**, there may be some confusion to this term, but at least per the @wabiverse (**Wabi Foundation**), this is the term we use to refer to open source software written upon the underlying [**ASWF landscape**](https://landscape.aswf.io), at it's core is Pixar's Universal Scene Description, which too, uses **TBB** as a dependency.

Currently, the **Wabi Foundation** is housing a mono-repo of all dependencies for this underlying landscape [here](https://github.com/wabiverse/MetaverseKit), since our first priority is to bring Pixar's **USD** to the **Swift** programming language, the goal is to have all dependencies of **USD** house their own `Package.swift` file at the root of their respective repository, as the **Swift Package Manager** is able to compile **Objective-C/C++, as well as C/C++, in addition to Swift**.

A good starting place to have a compliant `Package.swift` for **oneTBB** is this chunk of code [here](https://github.com/wabiverse/MetaverseKit/blob/main/Package.swift#L465-L477).

But I am more than happy to submit a `Package.swift` as a PR if this is something the developers of **oneTBB** would like to support.

Thank you.

**Comment by isaevil** [comment]
Hi @furby-tm. Thanks for you interest in contribution to oneTBB! We can accept your PR with support for Swift Package Manager, but the feature itself would be a community-supported though.

**Comment by furby-tm** [comment]
@isaevil I'm very happy to hear this! I am out of town but will issue a PR to support SwiftPM as soon as I'm back at my office. Thank you very much!

**Comment by arunparkugan** [comment]
@isaevil is this issue is still relevant?

**Comment by isaevil** [comment]
@arunparkugan I believe @furby-tm was planning to open a PR for that. @furby-tm is this still relevant?

**Comment by furby-tm** [comment]
@isaevil It's still relevant! Sorry for keeping this issue hanging for so long, Swift needs to incorporate [**this fix for Linux**](https://github.com/swiftlang/swift/pull/75662) before **oneTBB** can have it's own **`Package.swift`** manifest file.

Sorry for leaving this issue open for an eternity!

**Comment by wtholliday** [comment]
FWIW I'm also working on getting oneTBB into Swift Package Manager, but I'm taking the approach of a binary framework, built using this script: https://github.com/audulus/tbb-spm/blob/main/xcframework.sh

A pure Swift Package that doesn't use a binary is probably preferable though.

**Comment by furby-tm** [comment]
@wtholliday 
> FWIW I'm also working on getting oneTBB into Swift Package Manager, but I'm taking the approach of a binary framework, built using this script: https://github.com/audulus/tbb-spm/blob/main/xcframework.sh
> 
> A pure Swift Package that doesn't use a binary is probably preferable though.

Yeah, creating a binary `.xcframework` will only allow it to be utilized on **Apple** family devices.

I have **oneTBB** building across **Linux**, **Windows**, **macOS**, **visionOS**, and **iOS** [here](https://github.com/wabiverse/MetaverseKit/tree/main/Sources), it's the **OneTBB**, **TBBMalloc**, and **TBBMalloc**, and **MetaTBB** Swift targets specifically.

The goal is to clean it up / not have to do any directory re-structuring for a PR to this repo.

**Comment by wtholliday** [comment]
@furby-tm I'm trying to add a Package.swift in the `oneTBB` directory. Setting it up like this:

```
.target(
            name: "OneTBB",
            path: ".",
            sources: ["src/tbb/arena.cpp"],
            publicHeadersPath: "include",
            cxxSettings: [
                .define("_XOPEN_SOURCE", to: "1"),
                .define("__TBB_BUILD"),
                .define("TBB_NO_LIB_LINKAGE")
            ]
        ),
```

The `publicHeadersPath` can't be outside the `path`, so that forces path to be `.`. 

I get the error: ```
Multiple commands produce '.../Build/Products/Debug-iphoneos/tbb_OneTBB.bundle/en.lproj/MainMenu.nib'
```

So it seems SPM has inferred an executable of some sort. Seems harder to shoehorn this into SPM than build out a binary.

**Comment by furby-tm** [comment]
Yes, it's certainly a bit complicated to get things to line up, SwiftPM offers very little flexibility in the configuration. 

I've however not a clue how it's picked up a bundle, or what that nib is doing in there or any of that, the only way to create a bundle like that is by specifying a` resources`  path, which you do not appear to be doing?

Are you attempting to utilize Xcode for this? You should have much better results using the swift CLI in your terminal and just running swift build command.

**Comment by furby-tm** [comment]
@wtholliday ^ sorry forgot to ping you.

**Comment by wtholliday** [comment]
Thanks! Yeah it builds from the command line. I think Xcode sees a `src/tbb/main.cpp` and infers and executable. It's odd that SPM and Xcode treat swift packages so differently.

Anyway, I've got a `Package.swift` that at least builds now, though I suspect some files are missing and I'll get link errors.

```
// swift-tools-version:5.10
import PackageDescription

let package = Package(
    name: "tbb",
    defaultLocalization: "en",
    platforms: [
        .macOS(.v11), .iOS(.v13)
    ],
    products: [
        .library(
            name: "tbb",
            targets: ["OneTBB", "TBBMalloc", "TBBMallocProxy"])
    ],
    targets: [
        .target(
            name: "TBBMallocProxy",
            dependencies: [
                .target(name: "OneTBB"),
            ],
            path: ".",
            exclude: ["src/tbbmalloc_proxy/tbbmalloc_proxy.rc", "src/tbbmalloc_proxy/CMakeLists.txt"],
            sources: ["src/tbbmalloc_proxy"],
            publicHeadersPath: "include",
            cxxSettings: [
                .define("_XOPEN_SOURCE", to: "1"),
                .define("TBBPROXY_NO_DLLMAIN")
            ]
        ),

        .target(
            name: "TBBMalloc",
            dependencies: [
                .target(name: "OneTBB"),
                .target(name: "TBBMallocProxy"),
            ],
            path: ".",
            exclude: ["src/tbbmalloc/tbbmalloc.rc", "src/tbbmalloc/def", "src/tbbmalloc/CMakeLists.txt"],
            sources: ["src/tbbmalloc"],
            publicHeadersPath: ".",
            cxxSettings: [
                .define("_XOPEN_SOURCE", to: "1"),
                .define("__TBBMALLOC_BUILD", to: "1"),
                .define("TBBMALLOC_NO_DLLMAIN")
            ]
        ),

        .target(
            name: "OneTBB",
            path: ".",
            exclude: ["src/tbb/tbb.rc", "src/tbb/CMakeLists.txt"],
            sources: ["src/tbb"],
            publicHeadersPath: "include",
            cxxSettings: [
                .define("_XOPEN_SOURCE", to: "1"),
                .define("__TBB_BUILD"),
                .define("TBB_NO_LIB_LINKAGE")
            ]
        ),
    ],
    cxxLanguageStandard: .cxx20
)
```

--- 

Re SwiftBundler: looks awesome but this is a dealbreaker:

```
Swift Bundler supports automatic compilation of Metal shaders, however all Metal shader files must be siblings within a single directory and they cannot import any header files from the project.
```

And also when I've put Metal code inside a Swift Package, it breaks support for the GPU debugger/profiler.

**Comment by wtholliday** [comment]
Ok when I try to use it in my main Xcode project, I get the same errors about xib files. So this seems like a dealbreaker. Back to the binary framework approach.

**Comment by furby-tm** [comment]
> Ok when I try to use it in my main Xcode project, I get the same errors about xib files. So this seems like a dealbreaker. Back to the binary framework approach.

This is very strange, I'm unsure how this `.nib` problem keeps propagating for you, I've yet to see this strange behavior myself, how are you adding the **SwiftPM package** to your **Xcode Project**?

Are you using **Xcode's** &nbsp; **File > Add Package Dependency** &nbsp; mechanism for adding your **SwiftPM** package to your project?

Also, which version of **Xcode** are you using?

> Re SwiftBundler: looks awesome but this is a dealbreaker:
>
> _Swift Bundler supports automatic compilation of Metal shaders, however all Metal shader files must be siblings within a single directory and they cannot import any header files from the project._

Thanks for your feedback! I'll discuss this with @stackotter and see if we can make this work for you.

**Comment by furby-tm** [comment]
@wtholliday actually I just took a look at your code in your **SwiftPM** again and had realized you had the following line of code in there:
`defaultLocalization: "en"`, this was causing issues, more info on what that field does is [here](https://developer.apple.com/documentation/xcode/localizing-package-resources).

Simply removing that line from the OneTBB swift package should not cause any of the errors you were seeing, sorry that I didn't catch that at first glimpse!

**Comment by wtholliday** [comment]
@furby-tm unfortunately removing that line doesn't work. I get:

```
manifest property 'defaultLocalization' not set; it is required in the presence of localized resources
```

which is why I added the line

**Comment by furby-tm** [comment]
> @furby-tm unfortunately removing that line doesn't work. I get:
> 
> 
> 
> ```
> 
> manifest property 'defaultLocalization' not set; it is required in the presence of localized resources
> 
> ```
> 
> 
> 
> which is why I added the line

That is because you must remove your localized resources as well, [reference](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0278-package-manager-localized-resources.md#missing-default-localization).

**Comment by wtholliday** [comment]
@furby-tm ok I added "examples" to the excluded paths and that did it.
